src/ray/gcs/gcs_server/gcs_placement_group_manager.h

```diff
@@ -78,5 +78,88 @@ class GcsPlacementGroup {
   /// state of the gcs placement_group and so on (see gcs.proto).
   rpc::PlacementGroupTableData placement_group_table_data_;
 };
+
+using RegisterPlacementGroupCallback =
+    std::function<void(std::shared_ptr<GcsPlacementGroup>)>;
+/// GcsPlacementGroupManager is responsible for managing the lifecycle of all placement
+/// group. This class is not thread-safe.
+class GcsPlacementGroupManager : public rpc::PlacementGroupInfoHandler {
+ public:
+  /// Create a GcsPlacementGroupManager
+  ///
+  /// \param io_context The event loop to run the monitor on.
+  /// \param scheduler Used to schedule placement group creation tasks.
+  /// \param gcs_table_storage Used to flush placement group data to storage.
+  explicit GcsPlacementGroupManager(
+      boost::asio::io_context &io_context,
+      std::shared_ptr<GcsPlacementGroupSchedulerInterface> scheduler,
+      std::shared_ptr<gcs::GcsTableStorage> gcs_table_storage);
+  ~GcsPlacementGroupManager() = default;
+
+  /// Register placement_group asynchronously.
+  ///
+  /// \param request Contains the meta info to create the placement_group.
+  /// \param callback Will be invoked after the placement_group is created successfully or
+  /// be invoked immediately if the placement_group is already registered to
+  /// `registered_placement_groups_` and its state is `ALIVE`. The callback will not be
+  /// called in this case.
+  Status RegisterPlacementGroup(const rpc::CreatePlacementGroupRequest &request,
+                                RegisterPlacementGroupCallback callback);
+
+  /// Get the placement_group ID for the named placement_group. Returns nil if the
+  /// placement_group was not found.
+  /// \param name The name of the  placement_group to look up.
+  /// \returns PlacementGroupID The ID of the placement_group. Nil if the
+  /// placement_group was not found.
+  PlacementGroupID GetPlacementGroupIDByName(const std::string &name);
+
+  /// Schedule placement_groups in the `pending_placement_groups_` queue.
+  void SchedulePendingPlacementGroups();
```

Comment:
Should this be private?

Reply:
It will be hard to test if this function be priviate.

Reply:
Then, can you comment this is exposed for testing only? Or, use friend class to friend the test.

Reply:
> Then, can you comment this is exposed for testing only? Or, use friend class to friend the test.

The `SchedulePendingPlacementGroups` is used in many `TEST_F(GcsPlacementGroupSchedulerTest, XXX)`, so all of the test classes will be declared inside `GcsPlacementGroupManager`.
I think it is better to make `SchedulePendingPlacementGroups` private and virtual and then mock a GcsPlacementGroupManager which override `SchedulePendingPlacementGroups` and make it public.
I think this could be fixed in next PR :)